analyzers/src/SonarAnalyzer.CSharp/Rules/MutableFieldsShouldNotBe.cs

```diff
@@ -31,6 +31,8 @@ public abstract class MutableFieldsShouldNotBe : SonarDiagnosticAnalyzer
             ImmutableArray.Create(
                 KnownType.System_Collections_ObjectModel_ReadOnlyCollection_T,
                 KnownType.System_Collections_ObjectModel_ReadOnlyDictionary_TKey_TValue,
+                KnownType.System_Collections_Frozen_FrozenDictionary_TKey_TValue,
```

Comment:
https://sonarsource.github.io/rspec/#/rspec/S3887 seems to require an update too. While it doesn't list all the exceptions, as S2386 does, it gives an incomplete suggestion: 

> To fix this, you should either use an [Immutable collection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable?view=net-7.0) or remove the readonly field to clarify the behavior.

Immutable collections are only one possibility, `ReadOnly` and `Frozen` collections being equally valid alternatives.

Reply:
> This requires an RSPEC update

I have created SonarSource/rspec#3594 for both rules

Reply:
We can not check for `IReadOnlyCollection<T>` because it is implemented by a mixture of mutable and immutable collections:
[Microsoft.Extensions.Logging.Internal.FormattedLogValues](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.internal.formattedlogvalues?view=net-8.0)
[Microsoft.Extensions.Primitives.StringValues](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.primitives.stringvalues?view=net-8.0)
[System.ArraySegment<T>](https://learn.microsoft.com/en-us/dotnet/api/system.arraysegment-1?view=net-8.0)
[System.Collections.Concurrent.BlockingCollection<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1?view=net-8.0)
[System.Collections.Concurrent.ConcurrentBag<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1?view=net-8.0)
[System.Collections.Concurrent.ConcurrentDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-8.0)
[System.Collections.Concurrent.ConcurrentQueue<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentqueue-1?view=net-8.0)
[System.Collections.Concurrent.ConcurrentStack<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentstack-1?view=net-8.0)
[System.Collections.Frozen.FrozenDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.frozen.frozendictionary-2?view=net-8.0)
[System.Collections.Frozen.FrozenSet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.frozen.frozenset-1?view=net-8.0)
[System.Collections.Generic.Dictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-8.0)
[System.Collections.Generic.Dictionary<TKey,TValue>.KeyCollection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.keycollection?view=net-8.0)
[System.Collections.Generic.Dictionary<TKey,TValue>.ValueCollection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.valuecollection?view=net-8.0)
[System.Collections.Generic.HashSet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-8.0)
[System.Collections.Generic.IReadOnlyDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlydictionary-2?view=net-8.0)
[System.Collections.Generic.IReadOnlyList<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlylist-1?view=net-8.0)
[System.Collections.Generic.IReadOnlySet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlyset-1?view=net-8.0)
[System.Collections.Generic.LinkedList<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.linkedlist-1?view=net-8.0)
[System.Collections.Generic.List<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1?view=net-8.0)
[System.Collections.Generic.PriorityQueue<TElement,TPriority>.UnorderedItemsCollection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.priorityqueue-2.unordereditemscollection?view=net-8.0)
[System.Collections.Generic.Queue<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.queue-1?view=net-8.0)
[System.Collections.Generic.SortedDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.sorteddictionary-2?view=net-8.0)
[System.Collections.Generic.SortedDictionary<TKey,TValue>.KeyCollection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.sorteddictionary-2.keycollection?view=net-8.0)
[System.Collections.Generic.SortedDictionary<TKey,TValue>.ValueCollection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.sorteddictionary-2.valuecollection?view=net-8.0)
[System.Collections.Generic.SortedList<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.sortedlist-2?view=net-8.0)
[System.Collections.Generic.SortedSet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.sortedset-1?view=net-8.0)
[System.Collections.Generic.Stack<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.stack-1?view=net-8.0)
[System.Collections.Immutable.IImmutableDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.iimmutabledictionary-2?view=net-8.0)
[System.Collections.Immutable.IImmutableList<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.iimmutablelist-1?view=net-8.0)
[System.Collections.Immutable.IImmutableSet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.iimmutableset-1?view=net-8.0)
[System.Collections.Immutable.ImmutableArray<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1?view=net-8.0)
[System.Collections.Immutable.ImmutableArray<T>.Builder](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1.builder?view=net-8.0)
[System.Collections.Immutable.ImmutableDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutabledictionary-2?view=net-8.0)
[System.Collections.Immutable.ImmutableDictionary<TKey,TValue>.Builder](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutabledictionary-2.builder?view=net-8.0)
[System.Collections.Immutable.ImmutableHashSet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablehashset-1?view=net-8.0)
[System.Collections.Immutable.ImmutableHashSet<T>.Builder](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablehashset-1.builder?view=net-8.0)
[System.Collections.Immutable.ImmutableList<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablelist-1?view=net-8.0)
[System.Collections.Immutable.ImmutableList<T>.Builder](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablelist-1.builder?view=net-8.0)
[System.Collections.Immutable.ImmutableSortedDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablesorteddictionary-2?view=net-8.0)
[System.Collections.Immutable.ImmutableSortedDictionary<TKey,TValue>.Builder](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablesorteddictionary-2.builder?view=net-8.0)
[System.Collections.Immutable.ImmutableSortedSet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablesortedset-1?view=net-8.0)
[System.Collections.Immutable.ImmutableSortedSet<T>.Builder](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablesortedset-1.builder?view=net-8.0)
[System.Collections.ObjectModel.Collection<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.collection-1?view=net-8.0)
[System.Collections.ObjectModel.ReadOnlyCollection<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlycollection-1?view=net-8.0)
[System.Collections.ObjectModel.ReadOnlyDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlydictionary-2?view=net-8.0)
[System.Collections.ObjectModel.ReadOnlyDictionary<TKey,TValue>.KeyCollection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlydictionary-2.keycollection?view=net-8.0)
[System.Collections.ObjectModel.ReadOnlyDictionary<TKey,TValue>.ValueCollection](https://learn.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlydictionary-2.valuecollection?view=net-8.0)
[System.CommandLine.CompletionSourceList](https://learn.microsoft.com/en-us/dotnet/api/system.commandline.completionsourcelist?view=net-8.0)
[System.Diagnostics.TagList](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.taglist?view=net-8.0)
[System.Net.CookieCollection](https://learn.microsoft.com/en-us/dotnet/api/system.net.cookiecollection?view=net-8.0)
[System.Net.Http.Headers.HeaderStringValues](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.headerstringvalues?view=net-8.0)
[System.Net.Http.Headers.HttpHeadersNonValidated](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.headers.httpheadersnonvalidated?view=net-8.0)
[System.Net.Http.HttpRequestOptions](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httprequestoptions?view=net-8.0)
[System.Reflection.Metadata.AssemblyFileHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.assemblyfilehandlecollection?view=net-8.0)
[System.Reflection.Metadata.AssemblyReferenceHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.assemblyreferencehandlecollection?view=net-8.0)
[System.Reflection.Metadata.CustomAttributeHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.customattributehandlecollection?view=net-8.0)
[System.Reflection.Metadata.CustomDebugInformationHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.customdebuginformationhandlecollection?view=net-8.0)
[System.Reflection.Metadata.DeclarativeSecurityAttributeHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.declarativesecurityattributehandlecollection?view=net-8.0)
[System.Reflection.Metadata.DocumentHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.documenthandlecollection?view=net-8.0)
[System.Reflection.Metadata.EventDefinitionHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.eventdefinitionhandlecollection?view=net-8.0)
[System.Reflection.Metadata.ExportedTypeHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.exportedtypehandlecollection?view=net-8.0)
[System.Reflection.Metadata.FieldDefinitionHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.fielddefinitionhandlecollection?view=net-8.0)
[System.Reflection.Metadata.GenericParameterConstraintHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.genericparameterconstrainthandlecollection?view=net-8.0)
[System.Reflection.Metadata.GenericParameterHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.genericparameterhandlecollection?view=net-8.0)
[System.Reflection.Metadata.ImportScopeCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.importscopecollection?view=net-8.0)
[System.Reflection.Metadata.InterfaceImplementationHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.interfaceimplementationhandlecollection?view=net-8.0)
[System.Reflection.Metadata.LocalConstantHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.localconstanthandlecollection?view=net-8.0)
[System.Reflection.Metadata.LocalScopeHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.localscopehandlecollection?view=net-8.0)
[System.Reflection.Metadata.LocalVariableHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.localvariablehandlecollection?view=net-8.0)
[System.Reflection.Metadata.ManifestResourceHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.manifestresourcehandlecollection?view=net-8.0)
[System.Reflection.Metadata.MemberReferenceHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.memberreferencehandlecollection?view=net-8.0)
[System.Reflection.Metadata.MethodDebugInformationHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.methoddebuginformationhandlecollection?view=net-8.0)
[System.Reflection.Metadata.MethodDefinitionHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.methoddefinitionhandlecollection?view=net-8.0)
[System.Reflection.Metadata.MethodImplementationHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.methodimplementationhandlecollection?view=net-8.0)
[System.Reflection.Metadata.ParameterHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.parameterhandlecollection?view=net-8.0)
[System.Reflection.Metadata.PropertyDefinitionHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.propertydefinitionhandlecollection?view=net-8.0)
[System.Reflection.Metadata.TypeDefinitionHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.typedefinitionhandlecollection?view=net-8.0)
[System.Reflection.Metadata.TypeReferenceHandleCollection](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.metadata.typereferencehandlecollection?view=net-8.0)
[System.Security.Cryptography.Cose.CoseHeaderMap](https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.cose.coseheadermap?view=net-8.0)
[System.Text.RegularExpressions.CaptureCollection](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.capturecollection?view=net-8.0)
[System.Text.RegularExpressions.GroupCollection](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.groupcollection?view=net-8.0)
[System.Text.RegularExpressions.MatchCollection](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.matchcollection?view=net-8.0)

[Source](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlycollection-1?view=net-8.0)